pkg/xds/generator/secrets_generator.go

```diff
@@ -14,121 +16,106 @@ const OriginSecrets = "secrets"
-		caResources...,
-	)
-
-	return resources
 }
 
-func createDataPlaneProxySecrets(
+// GenerateSecretsFromTracker takes uses the SecretsTracker on Proxy and
+// generates whatever secrets were used in the config generation.
+func (g SecretsProxyGenerator) Generate(
 	ctx xds_context.Context,
 	proxy *core_xds.Proxy,
 ) (*core_xds.ResourceSet, error) {
-	mesh := ctx.Mesh.Resource
-
-	if !mesh.MTLSEnabled() {
-		return nil, nil
-	}
+	resources := core_xds.NewResourceSet()
 
-	var otherMeshes []*core_mesh.MeshResource
-	for _, otherMesh := range ctx.Mesh.Resources.CrossMeshGateways(mesh) {
-		if otherMesh.Mesh.GetMeta().GetName() != mesh.GetMeta().GetName() {
-			otherMeshes = append(otherMeshes, otherMesh.Mesh)
-		}
+	log := core.Log.WithName("secrets-generator")
+	switch {
+	case proxy.Dataplane != nil:
+		log = log.WithValues("proxy", proxy.Dataplane.GetMeta().GetName())
+	case proxy.ZoneIngressProxy != nil:
+		log = log.WithValues("zoneingress", proxy.ZoneIngress.GetMeta().GetName())
+	case proxy.ZoneEgressProxy != nil:
+		log = log.WithValues("zoneingress", proxy.ZoneEgressProxy.ZoneEgressResource.GetMeta().GetName())
 	}
 
-	identity, cas, err := ctx.ControlPlane.Secrets.GetForDataPlane(
-		proxy.Dataplane,
-		mesh,
-		otherMeshes,
-	)
-	if err != nil {
-		return nil, err
+	// We don't have a secrets tracker if we don't have a mesh (zone ingress/egress)
+	if proxy.SecretsTracker == nil {
+		return nil, nil
 	}
 
-	return CreateSecretResources(mesh, identity, cas), nil
-}
+	usedIdentity := proxy.SecretsTracker.UsedIdentity()
+	usedCas := proxy.SecretsTracker.UsedCas()
+	usedAllInOne := proxy.SecretsTracker.UsedAllInOne()
 
-func createZoneEgressSecrets(
-	ctx xds_context.Context,
-	proxy *core_xds.Proxy,
-) (*core_xds.ResourceSet, error) {
-	resources := core_xds.NewResourceSet()
-
-	for _, meshResources := range proxy.ZoneEgressProxy.MeshResourcesList {
-		if !meshResources.Mesh.MTLSEnabled() {
-			continue
+	if usedAllInOne {
+		otherMeshes := ctx.Mesh.Resources.OtherMeshes().Items
+		identity, allInOneCa, err := ctx.ControlPlane.Secrets.GetAllInOne(ctx.Mesh.Resource, proxy.Dataplane, otherMeshes)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to generate all in one CA")
 		}
 
-		if len(meshResources.ExternalServices) == 0 {
-			// https://github.com/envoyproxy/envoy/issues/9310
-			// Envoy has a behavior that if you include a secret over ADS that is not referenced anywhere
-			// all secrets are stuck in warming state.
-			// We need to only deliver secrets that are used in other parts of config.
-			// ZoneEgress only use identity and CA certs for external services.
-			// For internal services it just passes the traffic to ZoneIngress through SNI
-			continue
+		resources.Add(CreateCaSecretResource(proxy.SecretsTracker.RequestAllInOneCa().Name(), allInOneCa))
+		resources.Add(CreateIdentitySecretResource(proxy.SecretsTracker.RequestIdentityCert().Name(), identity))
+		log.V(1).Info("added all in one CA resources")
```

Comment:
should we return here? I'm not sure if there can be a case when we will go with this if and then with the second one